File: apps/desktop/src/main/diagnostics.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";

const SENSITIVE_PATTERN = /(token|secret|password|api[_-]?key|authorization|cookie)/gi;

export function diagnosticsLog(name: "electron" | "preload" | "renderer" | "runtime-api", message: string): void {
  const projectRoot = process.env.CLM_PROJECT_ROOT ?? process.cwd();
  const logDir = join(projectRoot, ".runtime", "logs");
  mkdirSync(logDir, { recursive: true });
  const clean = message.replace(SENSITIVE_PATTERN, "[redacted-key]");
  appendFileSync(join(logDir, `${name}.log`), `${new Date().toISOString()} ${clean}\n`, "utf8");
}

File: services/agent_runtime/app/core/diagnostics.py
from __future__ import annotations

import logging
import re
from pathlib import Path

from app.config import get_settings

SENSITIVE_PATTERN = re.compile(
    r"(token|secret|password|api[_-]?key|authorization|cookie|sqlite:///[^\\s]+|[A-Z]:\\\\[^\\r\\n\\t ]+)",
    re.IGNORECASE,
)


def redact_diagnostic(value: object) -> str:
    return SENSITIVE_PATTERN.sub("[redacted]", str(value))


def configure_runtime_logging() -> None:
    settings = get_settings()
    log_dir = settings.data_dir / "logs"
    log_dir.mkdir(parents=True, exist_ok=True)
    logging.basicConfig(
        filename=Path(log_dir / "runtime.log"),
        level=logging.INFO,
        format="%(asctime)s %(levelname)s %(name)s %(message)s",
        force=True,
    )
